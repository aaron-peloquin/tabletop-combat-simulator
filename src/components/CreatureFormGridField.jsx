import {
  Grid,
  TextField,
} from "@material-ui/core/"
import {withStyles} from "@material-ui/core/styles"
import {connect} from "react-redux"

/**
 * Builds <Grid> input fields for the <CreatureForm /> element
 * @param {obj} props
 *  @param {obj} classes MUI styles
 *  @param {str} dataKey the data key of this field
 *  @param {str} field the type of field
 *  @param {str} label the display label
 *  @param {str} size the grid width
 *  @param {func} onChange a callback function fired when this input field changes
 * @return {jsx} Component
 */
let CreatureFormGridField = (props) => {
  const {
    classes,
    dataKey,
    field,
    label,
    size,
    onChange,
  } = props
  let wrappedField; let DataField
  let dataValue = props.dataValue

  /**
   * Handle updating creature data, and call onChange callback
   * @param {obj} e event object
   * @param {str} key the key we are going to update
   * @return {void}
   */
  const updateData = (e, key) => {
    const curValue = e.target.value||""
    dataValue = curValue

    if (typeof onChange=="function") {
      onChange(key, curValue)
    }
  }

  const FieldProps = {
    "onChange": (e)=>{
      updateData(e, dataKey)
    },
    "className": classes.formField,
    "value": dataValue,
    "label": label,
  }

  switch (field) {
  default:
    DataField = <TextField {...FieldProps} />
    break
  case "multiline":
    DataField = <TextField multiline rows="3" {...FieldProps} />
    break
  }

  switch (size) {
  default:
  case "small":
    wrappedField = <Grid item xs={12} sm={4} md={4}>{DataField}</Grid>
    break
  case "medium":
    wrappedField = <Grid item xs={12} sm={6}>{DataField}</Grid>
    break
  case "large":
    wrappedField = <Grid item xs={12}>{DataField}</Grid>
    break
  }

  return wrappedField
}

const styles = (theme) => {
  return {
    formField: {
      width: "100%",
      marginTop: theme.spacing.unit,
      marginBottom: theme.spacing.unit,
      paddingLeft: theme.spacing.unit,
      paddingRight: theme.spacing.unit,
    },
  }
}

const MapStateToProps = (state, props) => {
  const dataValue = state.editing[props.dataKey]
  return {
    dataValue,
  }
}

CreatureFormGridField = withStyles(styles)(CreatureFormGridField)

export default connect(MapStateToProps)(CreatureFormGridField)
