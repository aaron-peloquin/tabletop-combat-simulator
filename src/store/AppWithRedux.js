import React from 'react'
import {initializeStore} from './store'

const isServer = typeof window === 'undefined'
const reduxStoreSymbol = Symbol("reduxStore")

function getOrCreateStore (initialState) {
  // Always make a new store if server, otherwise state is shared between requests
  if (isServer) {
    return initializeStore(initialState)
  }

  // Create store if unavailable on the client and set it on the window object
  if (!window[reduxStoreSymbol]) {
    window[reduxStoreSymbol] = initializeStore(initialState)
  }
  return window[reduxStoreSymbol]
}

export default (App) => {
  return class AppWithRedux extends React.Component {
    static async getInitialProps (appContext) {
      // Get or Create the store with `undefined` as initialState
      // This allows you to set a custom default initialState
      const reduxStore = getOrCreateStore()

      // Provide the store to getInitialProps of pages
      appContext.ctx.reduxStore = reduxStore

      let appProps = {}
      if (typeof App.getInitialProps === 'function') {
        appProps = await App.getInitialProps(appContext)
      }

      return {
        ...appProps,
        initialReduxState: reduxStore.getState()
      }
    }

    constructor (props) {
      super(props)
      this.reduxStore = getOrCreateStore(props.initialReduxState)
    }

    render () {
      return <App {...this.props} reduxStore={this.reduxStore} />
    }
  }
}
