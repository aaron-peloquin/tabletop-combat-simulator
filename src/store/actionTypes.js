/**
 * This is an object of Redux Actions. It mostly exists to
 *  ensure we do not create duplicate actions, but it can
 *  also be used as a list of our existing actions
*/
const actions = {
  CopyCreature: 'CopyCreature',
  DeleteAllCreatures: 'DeleteAllCreatures',
  DeleteCreature: 'DeleteCreature',
  SaveCreature: 'SaveCreature',
  SetEditCreature: 'SetEditCreature',
  ToggleSidebar: 'ToggleSidebar',
  RunSimulation: 'RunSimulation',
  UpdateEditCreature: 'UpdateEditCreature',
}

export default actions
