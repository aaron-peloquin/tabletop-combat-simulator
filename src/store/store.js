import {createStore, applyMiddleware, combineReducers} from "redux"
import {composeWithDevTools} from "redux-devtools-extension"
import thunkMiddleware from "redux-thunk"

/** Reducers */
import combat from "./reducer/combat"
import creatures from "./reducer/creatures"
import sideBar from "./reducer/sideBar"
import editing from "./reducer/editing"

/** Reducers */
const reducers = combineReducers({
  combat,
  creatures,
  editing,
  sideBar,
})

/**
 * Initializes the store
 * @param {obj} initialState the state our app starts with
 * @return {obj} the redux store
 */
export function initializeStore(initialState = initialState) {
  return createStore(reducers, initialState, composeWithDevTools(applyMiddleware(thunkMiddleware)))
}
