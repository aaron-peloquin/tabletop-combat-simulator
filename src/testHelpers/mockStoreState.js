export default {
  "creatures": [
    {
      "name": "Goplin",
      "cr": "1/4",
      "hash": "hivn3p0k",
      "description": "A stout monster",
      "defaultInitiative": 10,
      "armor": 11,
      "hp": 5,
      "hitDiceEquation": "1d20+1",
      "damageDiceEquation": "1d4"
    },
    {
      "name": "Elven Ranger",
      "cr": "1/2",
      "hash": "h3dxf5wl",
      "description": "A swift hunter",
      "defaultInitiative": 16,
      "armor": 13,
      "hp": 10,
      "hitDiceEquation": "1d20+8",
      "damageDiceEquation": "1d6+3"
    },
    {
      "name": "Knight",
      "cr": "1",
      "hash": "hee9p286",
      "description": "An honorable soldier",
      "defaultInitiative": 11,
      "armor": 18,
      "hp": 25,
      "hitDiceEquation": "1d20+4",
      "damageDiceEquation": "1d12+4"
    },
    {
      "name": "Lizardfolk",
      "cr": "3",
      "hash": "howx26rt",
      "description": "An ancient sneaky monster",
      "defaultInitiative": 14,
      "armor": 14,
      "hp": 15,
      "hitDiceEquation": "1d20+5",
      "damageDiceEquation": "1d10+2"
    }
  ],
  "sideBar": false,
}