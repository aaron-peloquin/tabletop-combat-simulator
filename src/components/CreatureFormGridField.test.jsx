import { Provider } from "react-redux"
import { createShallow } from "@material-ui/core/test-utils"
import toJson from "enzyme-to-json"
import { initializeStore } from "../store/store"

import CreatureFormGridField from "./CreatureFormGridField"
import mockStoreState from "../testHelpers/mockStoreState"

describe("<CreatureFormGridField />", ()=>{
  const RenderShallowUntilComponent = createShallow({"untilSelector":"CreatureFormGridField"})
  let store, props, Component
  beforeEach(async ()=>{
    store = initializeStore(mockStoreState)
    props = {
      onChange: jest.fn(()=>{}),
      creature: mockStoreState.creatures[0],
      dataKey: "name",
      label: "Name Field",
      size: "large",
    }
  })

  it("loads", () => {
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(typeof Component).toBe("object")
  })

  it("contains a <Grid> item", ()=>{
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(Component.find("WithStyles(Grid)[item=true]").length).toBe(1)
  })

  it("has a grid with xs of 12 (and no other MUI breakpoints) when large", ()=>{
    props.size = "large"
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(Component.find("WithStyles(Grid)[xs=12]").length).toBe(1)
    expect(Component.find("WithStyles(Grid)[sm]").length).toBe(0)
    expect(Component.find("WithStyles(Grid)[md]").length).toBe(0)
    expect(Component.find("WithStyles(Grid)[lg]").length).toBe(0)
    expect(Component.find("WithStyles(Grid)[xl]").length).toBe(0)
  })

  it("has a grid with any xs, sm, and md when small", ()=>{
    props.size = "small"
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(Component.find("WithStyles(Grid)[xs][sm][md]").length).toBe(1)
  })

  it("has a grid with any sm when medium", ()=>{
    props.size = "medium"
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(Component.find("WithStyles(Grid)[sm]").length).toBe(1)
  })

  it("can be a textarea", ()=>{
    props.field = "multiline"
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(Component.find("TextField[multiline=true]").length).toBe(1)
  })

  it("fires onChange callback", ()=>{
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    Component.find("TextField").simulate("change", {target:{value:"Some new name"}})
    expect(props.onChange).toHaveBeenCalledTimes(1)
  })

  it("updates props.creature on change", ()=> {
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    Component.find("TextField").simulate("change", {target:{value:"Some new name"}})
    expect(props.creature.name).toBe("Some new name")
  })

  it("snapshots", () => {
    Component = RenderShallowUntilComponent(<Provider store={store}><CreatureFormGridField {...props} /></Provider>)
    expect(toJson(Component)).toMatchSnapshot()
  })
})