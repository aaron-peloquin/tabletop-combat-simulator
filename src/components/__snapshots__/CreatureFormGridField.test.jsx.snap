// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CreatureFormGridField /> snapshots 1`] = `
<WithStyles(Grid)
  item={true}
  xs={12}
>
  <TextField
    className="CreatureFormGridField-formField-1"
    inputProps={
      Object {
        "data-cy": "name",
      }
    }
    label="Name Field"
    onChange={[Function]}
    required={false}
    select={false}
    value="Goplin"
    variant="standard"
  />
</WithStyles(Grid)>
`;
